### PR TITLE
[waspls] diagnostics for external imports and goto definition

### PR DESCRIPTION
**Pull request by craigmc08**
- Adds diagnostics for extimports trying to import JS code that does not exist
- Adds goto definition for extimports
- Refactors LSP to support a little bit of multithreading

### CODE REVIEW

File: waspc/packages/ts-inspect/README.md
Comment (infomiho) [inline waspc/packages/ts-inspect/README.md:12]: @craigmc08 I'm a bit confused about the API here.

There is a key called `filenames` but it only has a single string? Is this a mistake in writing the README?

File: waspc/packages/ts-inspect/src/index.ts
Comment (infomiho) [inline waspc/packages/ts-inspect/src/index.ts:26]: I would recommend defining a schema with Zod and parsing the input so you can be sure about the format 100% https://zod.dev/?id=basic-usage

File: waspc/packages/ts-inspect/src/exports.ts
Comment (infomiho) [inline waspc/packages/ts-inspect/src/exports.ts:40]: I would maybe extract the body of the loop into a helper function `getExportsForFile` and maybe do some `throw`ing and `catch`ing in the loop to output the errors and continue the loop instead of using `continue`

File: waspc/packages/ts-inspect/test/exportTests/complex.ts
Comment (infomiho) [inline waspc/packages/ts-inspect/test/exportTests/complex.ts:1]: What about exporting constants, do you differentiate those from functions?

```ts
export const name = "Craig";
```

File: waspc/packages/ts-inspect/README.md
Comment (craigmc08) [reply waspc/packages/ts-inspect/README.md]: That's a mistake in the README, should be an array. Fixed!

File: waspc/packages/ts-inspect/test/exportTests/complex.ts
Comment (craigmc08) [reply waspc/packages/ts-inspect/test/exportTests/complex.ts:1]: Exporting a constant is counted as a named export, it is not differentiated from functions. I added a new test case to document/check this behavior.

File: waspc/packages/ts-inspect/src/exports.ts
Comment (infomiho) [inline waspc/packages/ts-inspect/src/exports.ts:52]: @craigmc08 I think `continue` doesn't do anything here, the loop continues irregardless

File: waspc/waspls/test/Wasp/LSP/DebouncerTest.hs
Comment (infomiho) [inline waspc/waspls/test/Wasp/LSP/DebouncerTest.hs:19]: Does this mean that our unit tests have a minimum time of running of 20 seconds? 😄

File: waspc/packages/ts-inspect/src/exports.ts
Comment (craigmc08) [reply waspc/packages/ts-inspect/src/exports.ts]: Deleted `continue`

File: waspc/waspls/test/Wasp/LSP/DebouncerTest.hs
Comment (craigmc08) [reply waspc/waspls/test/Wasp/LSP/DebouncerTest.hs:19]: Thankfully not! `threadDelay` takes a delay in microseconds, so the longest test is only 40ms.

**Comment by infomiho** [comment]
Tested it locally, it feels really cool to be able to jump to definition from the Wasp file. I'll let @Martinsos do the review of the Haskell part.

- One thing that feels a bit weird: the popup with `w` or `g` above the definition (I'm guessing that's the first letter), can we avoid that?
- Another thing that feels weird is the fact I can hover over multiple things and click, but they take me to the same place. Maybe reduce the number of places for which we offer the goto definition feature if possible?

Uploading Screen Recording 2023-06-22 at 16.19.32.mov…

**Comment by craigmc08** [comment]
The single letter popups does seem like the first letter. I think that's because I respond with a single-character selection range in the TS file. I'll see if I can get the typescript compiler to give me more information on the definition.

The hover and click showing different underline areas -> I think that's because I don't specify what area the link covers, so VSCode assumes it is whatever word you're hovering over. Fixing this should be relatively quick.

**Review by Martinsos — APPROVED** [review]
@craigmc08  amazing job here, this is looking great!
I think all together it looks very sane and makes a lot of sense.

One big thing is rethinking the `Package` name, another is just ensuring that we keep LSP as understandable as possible -> it obviously got more complicated with these changes, which is to some degree unavoidable beacuse we have global state and threads, but if we can thighten it up even a bit more, that is a plus.

I would recommend following -> that we merge this PR, so we can get it into the Wasp 0.11, but that you also continue working on it immediately after it, and we then also merge the improvements relatively soon once they are ready.

I am not sure how to best handle this technically -> do we merge this PR, and you keep looking at it and we comment on these old comments on closed PR while you open a new one with changes?
Or do we create a branch from main called `main-for-1268`, have this PR point to it (while also merging it into `main`), and then continue working on it? Not sure, whatever you think is the best!

**Comment by craigmc08** [comment]
Ok I will merge this then.

Here's my plan for handling the merge while keeping the review comments: I'll merge this branch normally and create a new PR addressing feedback here + more code clean up. As I address the feedback with code changes, I'll quote comment your feedback on the modified code and link to those new comments in this PR. That way we have a way of checking that I responded to everything. Does that plan work for you?

**Comment by Martinsos** [comment]
> Ok I will merge this then.
> 
> Here's my plan for handling the merge while keeping the review comments: I'll merge this branch normally and create a new PR addressing feedback here + more code clean up. As I address the feedback with code changes, I'll quote comment your feedback on the modified code and link to those new comments in this PR. That way we have a way of checking that I responded to everything. Does that plan work for you?

Sure, go for it!

File: waspc/waspls/src/Wasp/LSP/Debouncer.hs
Comment (craigmc08) [reply waspc/waspls/src/Wasp/LSP/Debouncer.hs:44]: 🚀 https://github.com/wasp-lang/wasp/pull/1292/files#diff-d2463dff307c7af8f8dc13d40e1c035b53d940ed41dd80232f655dfa333bf8a3R52-R73

File: waspc/waspls/src/Wasp/LSP/Diagnostic.hs
Comment (craigmc08) [reply waspc/waspls/src/Wasp/LSP/Diagnostic.hs:25]: Good point! Renamed to `MissingExtImport`: https://github.com/wasp-lang/wasp/pull/1292/files#diff-0c7b958f2ee2116a36caf6e6a5509bd1d5befb21d309a669a22ed6a117b3e814R25

**Review by craigmc08 — COMMENTED** [review]
Responded to your feedback @Martinsos and gave links to the changes in the new draft PR: https://github.com/wasp-lang/wasp/pull/1292.

There are some comment threads in here that will take some discussion to resolve. Let's keep the old discussions in here, but start any new discussions ones in the new PR.

No rush on another round of review, I still have more of your feedback to work on that will take longer, just got most of the low-hanging fruit for now. I'll ping you when it's ready 😄

File: waspc/packages/ts-inspect/src/index.ts
Comment (Martinsos) [reply waspc/packages/ts-inspect/src/index.ts:3]: Whatever you like better! If what we have works for sure and we are not sure about fs.readFile(0), we can stick with what we have. Maaaaaybe you could rename it to `readAllFromStdin` -> but you will have a better feeling for this than me.

File: waspc/src/Wasp/Package.hs
Comment (Martinsos) [reply waspc/src/Wasp/Package.hs:47]: Ok all makes sense!
The thing is, naming is really bad on the `System.Process` side. `CreateProcess` -> pretty bad name, would be better if it was `ProcessCreator` or smth like that.
I forgot that that function is called `proc`, so seeing `getPackageProc` didn't help me much, I wasn't sure what `proc` their means. So I wouldn't stick to how they are calling it, instead would try to come up with a better name.

Ideas:
- `preparePackageForCalling` -> I like that it hides info about what it returns -> one can check that in the type. But, it also sounds like it does something every time, while it doesn't, it will do npm install and stuff only if needed, it doesn't do heavy prepartion every time.
- `ensurePackageIsPreparedForCalling` -> A bit better, `ensure` for me indicates it will ensure some conditions, which might need it will just confirm them, or actually produce them. But it sounds like it doesn't return anything. Which is not terrible, but still.
- `createPackageProcess` -> Plays on the idea that it returns `CreateProcess`. More haskellish, because if you give it two args, it inded becomes create process for package, but I think it also hides quite a bit so I don't like that a lot. `CReateProcess` name is giving us trouble here.
- `getPackageProcessCreator` -> Hides details about preparing the package, but makes it clear what it returns. Could be ok, caller doesn't really need to know how happens in the background. Maybe a better name would be `obtainPackageProcessCreator` -> `obtain` indicates it might need some heavy lifting, unlike `get` which usually means there is not much going on.

Looking at all this, maybe `preparePackageForCalling` or `obtainPackageProcessCreator` are the best in my opinion.

File: waspc/src/Wasp/TypeScript.hs
Comment (Martinsos) [reply waspc/src/Wasp/TypeScript.hs:88]: bump

File: waspc/src/Wasp/TypeScript.hs
Comment (Martinsos) [reply waspc/src/Wasp/TypeScript.hs:83]: bump

File: waspc/waspls/src/Wasp/LSP/Debouncer.hs
Comment (Martinsos) [reply waspc/waspls/src/Wasp/LSP/Debouncer.hs:30]: Maybe you could mention then that old event is cancelled, and that timer is reset.

File: waspc/waspls/src/Wasp/LSP/Debouncer.hs
Comment (Martinsos) [reply waspc/waspls/src/Wasp/LSP/Debouncer.hs:21]: Makes sense!

File: waspc/waspls/src/Wasp/LSP/ExtImport.hs
Comment (Martinsos) [reply waspc/waspls/src/Wasp/LSP/ExtImport.hs:1]: bump

File: waspc/waspls/src/Wasp/LSP/GotoDefinition.hs
Comment (Martinsos) [reply waspc/waspls/src/Wasp/LSP/GotoDefinition.hs:59]: bump

File: waspc/waspls/src/Wasp/LSP/ServerM.hs
Comment (Martinsos) [reply waspc/waspls/src/Wasp/LSP/ServerM.hs:79]: Yeah, maybe renaming then!

File: waspc/src/Wasp/Package.hs
Comment (craigmc08) [reply waspc/src/Wasp/Package.hs:47]: I renamed it to `getPackageProcessOptions`. I think there isn't much of a difference between get and obtain here, so I just chose the shorter one. And "process options" seems, to me, like a much better name than `CreateProcess`, because it literally is the options/settings that `createProcess` uses when it spawns a process.

File: waspc/waspls/src/Wasp/LSP/GotoDefinition.hs
Comment (craigmc08) [reply waspc/waspls/src/Wasp/LSP/GotoDefinition.hs:59]: It's a more convenient constructor for `LSP.LocationLink`. So I could rename it to `locationLink` but I'm not sure if that's more descriptive, since the type tells you that it returns a `LSP.LocationLink` anyways.

File: waspc/waspls/src/Wasp/LSP/Handlers.hs
Comment (craigmc08) [reply waspc/waspls/src/Wasp/LSP/Handlers.hs:1]: Moved the code that was cluttering this file to new files:
- Dynamic capability registering is in its own place now (DynamicHandlers).
- Doing analysis/diagnostics on the wasp file also has its own file now.
There's still a lot of comments on the dynamic capability stuff, but it just needs a lot of context from the LSP spec to explain what it's doing.

File: waspc/src/Wasp/Package.hs
Comment (craigmc08) [reply waspc/src/Wasp/Package.hs:4]: Renamed to `NodePackageFFI`. I didn't change how the variables are named inside the module, it all just refers to "package" since the module name gives the context.

> Are we ok with tying this just to Node?

I think tying it to node is better for now, especially since the code relies heavily on them being node programs. If we end up wanting to run code from other languages we can just rename the module again. #

File: waspc/packages/README.md
Comment (craigmc08) [reply waspc/packages/README.md:1]: I think since we went with `Wasp.NodePackageFFI`, it's ok to just keep this folder named `packages`. `js/npm/ts-packages` isn't much more descriptive.

**Pull request merged**
